Sources/SparkConnect/CaseInsensitiveDictionary.swift

```diff
@@ -0,0 +1,61 @@
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+import Foundation
+
+/// A  dictionary in which keys are case insensitive. The input dictionary can be
+/// accessed for cases where case-sensitive information is required.
+public struct CaseInsensitiveDictionary: Sendable {
+  public var originalDictionary: [String: Sendable]
+  private var keyLowerCasedDictionary: [String: Sendable] = [:]
+
+  init(_ originalDictionary: [String: Sendable] = [:]) {
+    self.originalDictionary = originalDictionary
+    for (key, value) in originalDictionary {
+      keyLowerCasedDictionary[key.lowercased()] = value
+    }
+  }
+
+  subscript(key: String) -> Sendable? {
+    get {
+      return keyLowerCasedDictionary[key.lowercased()]
+    }
+    set {
+      var newMap = originalDictionary.filter { $0.key.caseInsensitiveCompare(key) != .orderedSame }
```

Comment:
how about

```
    set {
      let lowerKey = key.lowercased()
      if let newValue = newValue {
        keyLowerCasedDictionary[lowerKey] = newValue
      } else {
        keyLowerCasedDictionary.removeValue(forKey: lowerKey)
      }
      originalDictionary = originalDictionary.filter { $0.key.lowercased() != lowerKey }
      if let newValue = newValue {
        originalDictionary[key] = newValue
      }
    }
```

Reply:
Actually, the original code pattern roughly follows Apache Spark code pattern which creates a new map.

https://github.com/apache/spark/blob/510edd763b912a21883e8a0dbaa3f661fedec234/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/CaseInsensitiveMap.scala#L41

Reply:
Let me update with yours.